File: app/src/main/java/com/owenlejeune/preffered/preferences/PreferredListPreference.java
package com.owenlejeune.preffered.preferences;

import android.content.Context;
import android.preference.ListPreference;
import android.util.AttributeSet;

import com.owenlejeune.preffered.Preferred;
import com.owenlejeune.preffered.listeners.PreferredChangeListener;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Implementation of {@link ListPreference} with the ability to add listeners for preference
 * changes, as well as auto-commit when the preference value changes
 */
@SuppressWarnings("unused")
public class PreferredListPreference extends ListPreference {

    private List<PreferredChangeListener> listeners;

    public PreferredListPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(attrs);
    }

    public PreferredListPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs);
    }

    public PreferredListPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    private void init(AttributeSet attrs) {
        if (Objects.isNull(getKey()) || getKey().isEmpty()) {
            throw new RuntimeException("No key defined for this preference.  Please set one in preferences.xml");
        }

        listeners = new ArrayList<>();

        refresh();
        setOnPreferenceChangeListener((p, o) -> setPreference(o));
    }

    private boolean setPreference(Object newVal) {
        boolean result = Preferred.putStringStatic(getKey(), (String)newVal);
        firePreferenceChanged(newVal);
        refresh();
        return result;
    }

    public String getPreference() {
        return Preferred.getStringStatic(getKey());
    }

    public void addPreferenceChangeListener(PreferredChangeListener listener) {
        listeners.add(listener);
    }

    public void removePreferenceChangeListener(PreferredChangeListener listener) {
        listeners.remove(listener);
    }

    private void firePreferenceChanged(final Object newValue) {
        listeners.forEach(l -> l.onPreferenceChange(newValue));
    }

    private void refresh() {
        setSummary(getEntry());
    }

}

File: library/src/main/java/com/owenlejeune/preffered/preferences/PreferredListPreference.kt
package com.owenlejeune.preffered.preferences

import android.content.Context
import androidx.preference.ListPreference
import android.util.AttributeSet
import com.owenlejeune.preffered.Preferred
import com.owenlejeune.preffered.listeners.PreferredChangeListener

/**
 * Implementation of [ListPreference] with the ability to add listeners for preference
 * changes, as well as auto-commit when the preference value changes
 */
open class PreferredListPreference @JvmOverloads constructor(context: Context,
                                                             attrs: AttributeSet? = null,
                                                             defStyleAttr: Int = 0)
    : ListPreference(context, attrs, defStyleAttr) {

    private lateinit var listeners: MutableList<PreferredChangeListener>

    init {
        key?.let {
            listeners = ArrayList()
            refresh()
            setOnPreferenceChangeListener { _, n ->  setPreference(n)}
            summary = if (getPreference().isNotBlank()) {
                entries[entryValues.indexOf(getPreference())]
            } else { entries[0] }
        }
        ?: throw RuntimeException("No key defined for this preference.  Please set one in preferences.xml")
    }

    private fun setPreference(newVal: Any): Boolean {
        val result = Preferred.putStringStatic(key, newVal as String)
        value = newVal
        firePreferenceChanged(newVal)
        refresh()
        return result
    }

    fun getPreference(): String = Preferred.getStringStatic(key)

    fun addPreferenceChangeListener(listener: PreferredChangeListener) {
        listeners.add(listener)
    }

    fun removePreferenceChangeListener(listener: PreferredChangeListener) {
        listeners.remove(listener)
    }

    private fun firePreferenceChanged(newVal: Any) {
        listeners.forEach { it.onPreferenceChange(newVal) }
    }

    private fun refresh() {
        summary = entry
    }

}